src/bitset.rs

```diff
@@ -295,6 +312,161 @@ impl<T> Iterator for Iter<T>
     }
 }
 
+/// This is similar to a `BitSet` but allows setting of value
+/// without unique ownership of the structure
+pub struct AtomicBitSet {
+    layer3: AtomicUsize,
+    layer2: Vec<AtomicUsize>,
+    layer1: Vec<AtomicUsize>,
+    layer0: Vec<AtomicUsize>,
+}
+
+impl AtomicBitSet {
+    /// Creates an empty `BitSet`.
+    pub fn new() -> AtomicBitSet {
+        AtomicBitSet {
+            layer3: AtomicUsize::new(0),
+            layer2: Vec::new(),
+            layer1: Vec::new(),
+            layer0: Vec::new(),
+        }
+    }
+
+    #[inline]
+    fn valid_range(max: Index) {
+        if (MAX_EID as u32) < max {
+            panic!("Expected index to be less then {}, found {}", MAX_EID, max);
+        }
+    }
+
+    /// Creates an empty `BitSet`, preallocated for up to `max` indices.
+    pub fn with_capacity(max: Index) -> AtomicBitSet {
+        Self::valid_range(max);
+        let mut value = AtomicBitSet::new();
+        value.extend(max);
+        value
+    }
+
+    #[inline(never)]
+    fn extend(&mut self, id: Index) {
+        Self::valid_range(id);
+        let (p0, p1, p2) = offsets(id);
+
+        if self.layer2.len() <= p2 {
+            let count = p2 - self.layer2.len() + 1;
+            self.layer2.extend(repeat(0).map(|_| AtomicUsize::new(0)).take(count));
+        }
+        if self.layer1.len() <= p1 {
+            let count = p1 - self.layer1.len() + 1;
+            self.layer1.extend(repeat(0).map(|_| AtomicUsize::new(0)).take(count));
+        }
+        if self.layer0.len() <= p0 {
+            let count = p0 - self.layer0.len() + 1;
+            self.layer0.extend(repeat(0).map(|_| AtomicUsize::new(0)).take(count));
+        }
+    }
+
+    /// Adds `id` to the `AtomicBitSet`. Returns `true` if the value was
+    /// already in the set.
+    ///
+    /// Because we cannot safely extend an AtomicBitSet without unique ownership
+    /// this will panic if the Index is out of range
+    #[inline]
+    pub fn add_atomic(&self, id: Index) -> bool {
+        let (p0, p1, p2) = offsets(id);
+        let old = self.layer0[p0].fetch_or(id.mask(SHIFT0), Ordering::Relaxed);
+        if old & id.mask(SHIFT0) != 0 {
+            return true;
+        }
+
+        self.layer1[p1].fetch_or(id.mask(SHIFT1), Ordering::Relaxed);
+        self.layer2[p2].fetch_or(id.mask(SHIFT2), Ordering::Relaxed);
+        self.layer3.fetch_or(id.mask(SHIFT3), Ordering::Relaxed);
+        false
+    }
+
+    /// Adds `id` to the `BitSet`. Returns `true` if the value was
+    /// already in the set.
+    #[inline]
+    pub fn add(&mut self, id: Index) -> bool {
+        if id.offset(SHIFT1) >= self.layer0.len() {
+            self.extend(id);
+        }
+
+        self.add_atomic(id)
+    }
+
+    /// Removes `id` from the set, returns `true` if the value
+    /// was removed, and `false` if the value was not set
+    /// to begin with.
+    #[inline]
+    pub fn remove(&mut self, id: Index) -> bool {
```

Comment:
So it is sound either atomically add or atomically delete, but not both. If it was a plane bitset it would be safe to do both and this could be `&self`.

But the hierarchy causes problems. The atomic add will set the layers above it, and the worst case is that two writers both set the same bit, which has no effect.

In the case of an atomic remove, whoever migrates the layer to 0 is required to clear the bit in the layer above. This is also sound, as long as the other writers are doing the same.

The issue is when these two are mixed. If Thread 1 sets a bit, then sets the bits in the layers above. And Thread 2 removes the same bit there is a race between the two threads that will cause the layers above to enter an inconsistent state.

So therefore, the only way I can make sure this is sound is to make sure the removes require unique ownership, which prevents the case when thread A is setting and Thread B is clearing.